### PR TITLE
Copy between linear data and texture operation tests

### PR DESCRIPTION
**Pull request by ponitka**
PTAL

### CODE REVIEW

**Comment by kainino0x** [comment]
I can't get to this today, sorry :(

**Comment by ponitka** [comment]
> I can't get to this today, sorry :(

No problem!

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+  ): void {
+    const { mipSize, bytesPerRow, rowsPerImage, byteLength } = fullTextureCopyLayout;
+    const { dst, begin, end } = this.createAlignedCopyForMapRead(fullData, byteLength, 0);
+
+    // We add an eventual async expectation which will update the full data and then add
+    // other eventual async expectations to ensure it will be correct.
+    this.eventualAsyncExpectation(async () => {
+      await dst.mapAsync(GPUMapMode.READ);
+      const actual = new Uint8Array(dst.getMappedRange()).slice(begin, end);
+      this.updateFullData(
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:268]: updateFullData could be done before entering the eventual expectation. In general we should minimize the amount of work done in eventual expectations, and check against known data as much as possible.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+    dimension = '2d',
+    initMethod,
+    checkMethod,
+  }: {
+    textureDataLayout: GPUTextureDataLayout;
+    copySize: GPUExtent3DDict;
+    dataSize: number;
+    mipLevel?: number;
+    origin?: Required<GPUOrigin3DDict>;
+    textureSize: [number, number, number];
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:305]: nit Extent3D?

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+    format: SizedTextureFormat;
+    dimension?: GPUTextureDimension;
+    initMethod: string;
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:308]: nit: make an enum, ditto for checkMethod

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+        this.device.defaultQueue.submit([encoder.finish()]);
+
+        break;
+      }
+      default:
+        unreachable();
+    }
+  }
+
+  checkData(
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:167]: nit: It could help readability to have comments explaining what these functions do, and highlight the different between checkData and fullCheck, ditto for other largish functions.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 5]
```diff
@@ -0,0 +1,742 @@
...
+  }
+}
+
+// This is a helper function used for filtering test parameters
+// TODO: Modify this after introducing tests with rendering.
+function formatCanBeTested({ format }: { format: SizedTextureFormat }): boolean {
+  return kSizedTextureFormatInfo[format].copyDst && kSizedTextureFormatInfo[format].copySrc;
+}
+
+const kMethodsToTest = [
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:374]: We should comment why this gives us good coverage.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 6]
```diff
@@ -0,0 +1,742 @@
...
+    params()
+      .combine(kMethodsToTest)
+      .combine([
+        { bytesPerRowPadding: 0, rowsPerImagePaddingInBlocks: 0 }, // no padding
+        { bytesPerRowPadding: 0, rowsPerImagePaddingInBlocks: 6 }, // rowsPerImage padding
+        { bytesPerRowPadding: 6, rowsPerImagePaddingInBlocks: 0 }, // bytesPerRow padding
+        { bytesPerRowPadding: 15, rowsPerImagePaddingInBlocks: 17 }, // both paddings
+      ])
+      .combine([
+        { copyWidthInBlocks: 3, copyHeightInBlocks: 4, copyDepth: 5 }, // standard copy, covers bytesPerRow = 256 and copyDepth % 2 == 1.
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:400]: nit: explains that this gives coverage of special cases in the D3D12 backend.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+  const textureSize = [
+    (texturePhysicalSizeAtMipLevelInBlocks.width * info.blockWidth!) << mipLevel,
+    (texturePhysicalSizeAtMipLevelInBlocks.height * info.blockHeight!) << mipLevel,
+    texturePhysicalSizeAtMipLevelInBlocks.depth,
+  ];
+  yield {
+    textureSize: textureSize as [number, number, number],
+  };
+
+  // We choose width and height of the texture so that the values are divisible by blockWidth and
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:621]: The logic here is pretty hard to follow, why not just do a width-1 and height-1?

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 8]
```diff
@@ -0,0 +1,742 @@
+export const description = `writeTexture + copyBufferToTexture + copyTextureToBuffer operation tests.
+
+* copy_with_various_rows_per_image_and_bytes_per_row: covers
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:3]: This test plan is a bit dry, can we explain with more prose what happens? Often this would mean moving the large comment that's before the tests here, and explains the special cases in a bit more detail.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 9]
```diff
@@ -0,0 +1,742 @@
...
+//    bufferSize - offset < bytesPerImage * copyExtent.depth and copyExtent needs to be clamped for all BC formats
+g.test('copy_various_mip_levels')
+  .params(
+    params()
+      .combine(kMethodsToTest)
+      .combine([
+        {
+          copySizeInBlocks: { width: 5, height: 4, depth: 1 },
+          originInBlocks: { x: 3, y: 2, z: 0 },
+          texturePhysicalSizeAtMipLevelInBlocks: { width: 8, height: 6, depth: 1 },
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:665]: nit: IIUC it looks like this could be a private param (`_texturePhysicalSizeAtMipLevelInBlocks`) since the actual identifying part of the params is `textureSize`

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 10]
```diff
@@ -0,0 +1,742 @@
...
+  format: SizedTextureFormat;
+  mipLevel: number;
+  texturePhysicalSizeAtMipLevelInBlocks: GPUExtent3DDict;
+}): Generator<{ textureSize: [number, number, number] }> {
+  const info = kSizedTextureFormatInfo[format];
+  const textureSize = [
+    (texturePhysicalSizeAtMipLevelInBlocks.width * info.blockWidth!) << mipLevel,
+    (texturePhysicalSizeAtMipLevelInBlocks.height * info.blockHeight!) << mipLevel,
+    texturePhysicalSizeAtMipLevelInBlocks.depth,
+  ];
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:616]: nit: `as const` here means you don't need `as [number, number, number]` below.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 11]
```diff
@@ -0,0 +1,742 @@
...
+  // blockHeight respectively and so that the virtual size at mip level corresponds to the same
+  // physical size.
+  // Virtual size at mip level with modified width has width = (physical size width) - (blockWidth / 2).
+  // Virtual size at mip level with modified height has height = (physical size height) - (blockHeight / 2).
+  assert(mipLevel > 0);
+  assert(textureSize[0] > 0 && textureSize[1] > 0);
+  const modifiedWidth = textureSize[0] - (info.blockWidth! << (mipLevel - 1));
+  const modifiedHeight = textureSize[1] - (info.blockHeight! << (mipLevel - 1));
+
+  if (info.blockWidth! > 1 && modifiedWidth !== textureSize[0]) {
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:631]: nit: factor out into `modifyWidth` and `modifyHeight` and then use `modifyWidth && modifyHeight` below

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 12]
```diff
@@ -0,0 +1,742 @@
...
+  };
+
+  // We choose width and height of the texture so that the values are divisible by blockWidth and
+  // blockHeight respectively and so that the virtual size at mip level corresponds to the same
+  // physical size.
+  // Virtual size at mip level with modified width has width = (physical size width) - (blockWidth / 2).
+  // Virtual size at mip level with modified height has height = (physical size height) - (blockHeight / 2).
+  assert(mipLevel > 0);
+  assert(textureSize[0] > 0 && textureSize[1] > 0);
+  const modifiedWidth = textureSize[0] - (info.blockWidth! << (mipLevel - 1));
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:628]: easier to read?:
```ts
  // above:
  const widthAtThisLevel = texturePhysicalSizeAtMipLevelInBlocks.width * info.blockWidth;
  // here:
  const widthAtPrevLevel = widthAtThisLevel << 1;
  const modifiedWidth = (widthAtPrevLevel - 1) << (mipLevel - 1);
```

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 13]
```diff
@@ -0,0 +1,742 @@
...
+    yield {
+      textureSize: [modifiedWidth, modifiedHeight, textureSize[2]],
+    };
+  }
+}
+
+// Test that copying various mip levels works.
+// Covers two special code paths:
+//    the physical size of the subresouce is not equal to the logical size
+//    bufferSize - offset < bytesPerImage * copyExtent.depth and copyExtent needs to be clamped for all BC formats
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:656]: nit:
```suggestion
//    bufferSize - offset < bytesPerImage * copyExtent.depth and copyExtent needs to be clamped for all block formats
```

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 6 above)
Comment (ponitka) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: Done

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 8 above)
Comment (ponitka) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: Done

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 7 above)
Comment (Kangz) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:781]: Still working on this re-review, will take a look at this function and the last two commits again tomorrow.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 10 above)
Comment (kainino0x) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:776]: For some reason I couldn't find the mutating code, but you found exactly the backup solution I was going to suggest :)

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 12 above)
Comment (kainino0x) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: aha, nice.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 14]
```diff
@@ -0,0 +1,963 @@
...
+        layout.bytesPerRow * (copyExtent.height / kSizedTextureFormatInfo[format].blockHeight - 1) +
+        (copyExtent.width / kSizedTextureFormatInfo[format].blockWidth) *
+          kSizedTextureFormatInfo[format].bytesPerBlock;
+      return bytesPerImage * (copyExtent.depth - 1) + bytesInLastSlice;
+    }
+  }
+
+  // Offset for a particular texel in the linear texture data
+  getTexelOffsetInBytes(
+    textureDataLayout: GPUTextureDataLayout,
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:104]: Could this be `Required<GPUTextureDataLayout>`?

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 15]
```diff
@@ -0,0 +1,963 @@
...
+      appliedCopyView,
+      { buffer, ...appliedDataLayout },
+      appliedCheckSize
+    );
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    for (const texel of this.iterateBlockRows(checkSize, origin, format)) {
+      const rowOffset = this.getTexelOffsetInBytes(textureDataLayout, format, texel, origin);
+      const rowLength = this.bytesInACompleteRow(checkSize.width, format);
+      this.expectContents(buffer, expected.slice(rowOffset, rowOffset + rowLength), rowOffset);
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:329]: We should expect something about the data in between rows and in between images, too, although I don't know what it is. (Probably that it's zeroed? Maybe that it's preserved?)

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 16]
```diff
@@ -0,0 +1,963 @@
...
+    }
+  }
+
+  // Used for fullCheck.
+  // Copies the whole texture into linear data stored in a buffer for further checks.
+  getFullData(
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:335]: copyWholeTextureToNewBuffer

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 17]
```diff
@@ -0,0 +1,963 @@
...
+      }
+      default:
+        unreachable();
+    }
+  }
+
+  // We check an appropriate part of the texture against the given data.
+  // Used directly with PartialCopyT2B check method (for a subpart of the texture)
+  // and by fullCheck function with FullCopyT2B check method (for the whole texture).
+  checkData(
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:292]: copyPartialTextureToBufferAndCheckContents

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 18]
```diff
@@ -0,0 +1,963 @@
...
+interface TextureCopyViewWithRequiredOrigin {
+  texture: GPUTexture;
+  mipLevel: number | undefined;
+  origin: Required<GPUOrigin3DDict>;
+}
+
+type InitMethod = 'WriteTexture' | 'CopyB2T';
+type CheckMethod = 'PartialCopyT2B' | 'FullCopyT2B';
+
+// This describes in what form the arguments will be passed to WriteTexture/CopyB2T/CopyT2B.
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:54]: fyi: If you use jsdoc comments:
```ts
/** comment */
or
/**
 * comment
 */
```
then this will be picked up by IDEs and shown on mouseover.

Not super necessary within a file, but useful for helpers in shared files. We don't use this much in this project yet, but a few things have been updated to use it.

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 19]
```diff
@@ -0,0 +1,963 @@
...
+        { width: mipSize[0], height: mipSize[1], depth: mipSize[2] },
+        actual
+      );
+      dst.destroy();
+    });
+  }
+
+  // Tests copy between linear data and texture by creating a texture, putting some data into it with WriteTexture/CopyB2T,
+  // then getting data for the whole texture/for a part of it back and comparing it with the expectation.
+  testRun({
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:434]: uploadTextureAndVerifyCopy

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 20]
```diff
@@ -0,0 +1,963 @@
...
+        fullDataOffset
+      );
+    }
+  }
+
+  // Used for checking whether the whole texture was updated correctly by initTexture.
+  // Takes fullData returned by getFullData before the copy operation which is the original texture data,
+  // then updates it with updateFullData and checks the texture against the updated data after the
+  // copy operation.
+  fullCheck(
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:395]: copyWholeTextureToBufferAndCheckContents

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 21]
```diff
@@ -0,0 +1,963 @@
...
+      mipSize
+    );
+    this.device.defaultQueue.submit([encoder.finish()]);
+
+    return buffer;
+  }
+
+  // Takes the data returned by getFullData and updates it after a copy operation on the texture
+  // by emulating the copy behaviour here directly.
+  updateFullData(
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:358]: updateLinearTextureDataSubRectangle

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 22]
```diff
@@ -0,0 +1,963 @@
...
+    fullTextureCopyLayout: TextureCopyLayout,
+    texturePartialDataLayout: GPUTextureDataLayout,
+    copySize: GPUExtent3DDict,
+    origin: Required<GPUOrigin3DDict>,
+    format: SizedTextureFormat,
+    fullData: Uint8Array,
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:364]: destination, destinationCopyLayout

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 23]
```diff
@@ -0,0 +1,963 @@
...
+    partialData: Uint8Array
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:365]: source, sourceDataLayout

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 24]
```diff
@@ -0,0 +1,963 @@
...
+  ): GPUExtent3D {
+    if (changeBeforePass === 'arrays') {
+      return [width, height, depth];
+    } else {
+      return { width, height, depth };
+    }
+  }
+
+  // Put data into a part of the texture with an appropriate method.
+  initTexture(
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:232]: uploadLinearTextureDataToTextureSubRectangle

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 25]
```diff
@@ -239,15 +263,18 @@ class CopyBetweenLinearDataAndTextureTest extends GPUTest {
...
       );
-      const rowLength =
-        (copySize.width / kSizedTextureFormatInfo[format].blockWidth!) *
-        kSizedTextureFormatInfo[format].bytesPerBlock!;
-      for (let b = 0; b < rowLength; ++b) {
-        fullData[fullDataOffset + b] = partialData[partialDataOffset + b];
-      }
+      const rowLength = this.bytesInACompleteRow(copySize.width, format);
+      fullData.set(
+        partialData.slice(partialDataOffset, partialDataOffset + rowLength),
```
[the remark this answers is not in the document]
Comment (ponitka) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: Done

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 24 above)
Comment (Kangz) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: SubBox?

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 26]
```diff
@@ -144,20 +147,90 @@ class CopyBetweenLinearDataAndTextureTest extends GPUTest {
     return arr;
   }
 
+  // This is used for testing passing undefined members of TextureDataLayout instead of actual values
+  // where possible. Passing arguments as values and not as objects so that they are passed by copy
+  // and not by reference.
+  appliedDataLayout(
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:167]: nit: undefDataLayoutIfNeeded?

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 27]
```diff
@@ -144,20 +147,90 @@ class CopyBetweenLinearDataAndTextureTest extends GPUTest {
...
+        rowsPerImage = undefined;
+      }
+    }
+    return { offset, bytesPerRow, rowsPerImage };
+  }
+
+  // This is used for testing passing undefined members of TextureCopyView instead of actual values
+  // where possible. Passing arguments as values and not as objects so that they are passed by copy
+  // and not by reference.
+  appliedCopyView(
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:187]: ditto

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 28]
```diff
@@ -771,3 +863,64 @@ g.test('copy_various_mip_levels')
...
+          rowsPerImage: 0,
+          copySize: { width: 100, height: 1, depth: 1 },
+          origin: { x: 0, y: 0, z: 0 },
+        }, // copying one row: bytesPerRow can be < bytesInACompleteRow = 400
+        {
+          bytesPerRow: 256,
+          rowsPerImage: 0,
+          copySize: { width: 3, height: 3, depth: 1 },
+          origin: { x: 0, y: 0, z: 0 },
+        }, // copying one slice: rowsPerImage = 0 will be set to undefined.
```
Comment (Kangz) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: Can we also test that rowsPerImage = 2 is allowed and works even if less than copySize.depth?

**Review by Kangz — APPROVED** [review]
LGTM!

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts (hunk 15 above)
Comment (ponitka) [reply src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts]: I added validation that the other values are preserved.

**Review by ponitka — COMMENTED** [review]
> Maybe it would be easier to always `validateOtherBytesInBuffer` so the code below can become:
> 
> ```js
>     const expectedData = originalBufferData
>     for (const texel of this.iterateBlockRows(checkSize, origin, format)) {
>        memcpy the row in the original data
>     }
> 
>    this.expectContents(buffer, expectedData);
> ```

Seems better. Done

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 29]
```diff
@@ -0,0 +1,1048 @@
...
+/** This describes in what form the arguments will be passed to WriteTexture/CopyB2T/CopyT2B.
+ * If undefined, then default values are passed as undefined instead of default values.
+ * If arrays, then GPUOrigin3D and GPUExtent3D are passed as [number, number, number]. */
+type ChangeBeforePass = 'none' | 'undefined' | 'arrays';
+
+/** Each combination of methods can assume that the ones before it were tested and work correctly. */
+const kMethodsToTest = [
+  /** We make sure that CopyT2B works when copying the whole texture for renderable formats:
+   * TODO */
+  /** Then we make sure that WriteTexture works for all formats: */
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:68]: fyi I don't think jsdoc-style comments have any effect on values/expressions like here, or on `g.test()` below. They just get treated as regular comments (which is fine though a little more verbose).

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 30]
```diff
@@ -0,0 +1,1048 @@
...
+      .combine([
+        { offsetInBlocks: 0, dataPaddingInBytes: 0 } /** no offset and no padding */,
+        { offsetInBlocks: 1, dataPaddingInBytes: 0 } /** offset = 1 */,
+        { offsetInBlocks: 2, dataPaddingInBytes: 0 } /** offset = 2 */,
+        { offsetInBlocks: 15, dataPaddingInBytes: 0 } /** offset = 15 */,
+        { offsetInBlocks: 16, dataPaddingInBytes: 0 } /** offset = 16 */,
+        {
+          offsetInBlocks: 242,
+          dataPaddingInBytes: 0,
+        } /** for rgba8unorm format: offset + bytesInCopyExtentPerRow = 242 + 12 = 256 = bytesPerRow */,
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:692]: nit: The jsdoc style (even though it's ignored) implies that the comment applies to the thing after it, which isn't the case here

**Review by Kangz — APPROVED** [review]
LGTM, no additional comments

File: src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts [hunk 31]
```diff
@@ -0,0 +1,1048 @@
...
+  const textureSize: [number, number, number] = [
+    widthAtThisLevel << mipLevel,
+    heightAtThisLevel << mipLevel,
+    _texturePhysicalSizeAtMipLevelInBlocks.depth,
+  ];
+  yield {
+    textureSize,
+  };
+
+  /** We choose width and height of the texture so that the values are divisible by blockWidth and
```
Comment (kainino0x) [inline src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts:856]: Also this doc comment is attached to the `const widthAtPrevLevel`, should just be a regular comment so it doesn't do that

**Comment by ponitka** [comment]
Thanks for fixing the comment issues @kainino0x!
